### PR TITLE
Ukhas parser module

### PR DESCRIPTION
**Pull request by adamgreig**
**Opening this up for discussion (not ready to be pulled yet).**

I've started out on writing some kind of parser for the UKHAS protocol. So far I've implemented the various checksumming algorithms we've used in utils/checksum, but there could be discussion as to whether they should be in a class, and whether we should even support Fletcher, or if we should build in support for more algorithms now.

For the actual parser, it is a class that takes a config object on initialisation which I envisage will be a list of payload configuration objects. It might be preferable to load this once the payload name is determined, instead, which will require something with access to the couch db (payload configurations are likely to be stored as their own documents, after all). The configuration specifies everything useful the current XML specifies, but there is plenty of scope for discussion as to how it does this and whether we should include anything else (or remove anything I've currently listed). My ideas for the config can be found in the docstring for the ukhas_parser.py file. In particular, it might be worth considering allowing the radio, telemetry and sentence types to be lists, corresponding to payloads with multiple transmission methods. It may even be worth letting other values be lists, for instance baud rate to allow quick selecting multiple bauds.

The parser currently has a pre_parse and an (empty) parse method. pre_parse does some guesswork to try and see if this is a UKHAS protocol - in particular, looks for $$ at the start, \* either third or fifth from the end, and a non-zero sized list when the string is split by comma. If any of the above fail, it returns None - this is also open to debate. If they all succeed, it extracts the payload from the first position in the telemetry and returns it.

To summarise points in need of discussion:
- Should checksums go in a class, or are they okay as they are?
- Should we bother with a Fletcher implementation?
- Should we write support for even more algorithms now?
- Should parser modules be given config lists, or some interface to a couchdb interface they can later poll?
- How should that configuration dictionary look?
- Are pre_parse and parse suitable methods?
- What should pre_parse return when it decides it cannot parse the string? What about when it decides it can?

Any additional commits I push to randomskk/ukhas-parser-module will show up in this pull request and become visible on the "Files Changed" tab, so it should be fairly easy to keep track of what the parser module looks like as it develops.

### CODE REVIEW

**Comment by danielrichman** [comment]
- Checksums are fine where they are in my opinion
- Wrt implementing algorithms (both issues) we need to consider if they will actually be used and/or if we want to encourage their use. I'm not sure we need more algos unless you think there's a problem with crc16? most people use that or xor. We would have to double check and find out who used Fletcher and if they still use it.
- I think there's an argument for giving parse() a dict for the payload only; that is, pre_parse identifies what payload it is and if it's the correct protocol/whatever, the parser looks up which filters it has to run for that payload, does that, looks up the config for that payload, and gives it to parse(). This would avoid a lot of duplication; if every parser module had to go searching through a config or couch object it could get messy. This also means we can move the config around w/o affecting the parser modules.
- pre_parse and parse sound fine. I think that pre_parse should either return None or the payload name as a child of basestring, depending on whether it can or can't parse it.

configuration dict: referring to the Google doc on implementing habitat:
- Mission Control - stored as a/several CouchDB documents?
  - for each parser module
    - list of callsigns that use it - or wildcard enable
    - optional configuration for each callsign (like current XML)
    - list of optional filters/hotfixes, and when they should be executed
  - for each “flight”
    - list of callsigns involved
    - time range for the flight

Therefore I suggest that the config doc looks something like this, which I've based off your example in the docstring. It's essentially the same, but:

```
"configuration":
{
  "flights":
  {
    123:  /* id for life */
    {
      "callsigns":
      {
        'a1':
        {
          /* metadata irrelevant to implementation of habitat; radio, rtty settings, etc. */
        }
        'habitat':
        {
          /* ... */
        }
      },
      "start-time": timestamp,
      "end-time": timestamp
      /* ... metadata incl, who's flight it is, contact info, launch site, etc... */
    }
  },
  "parser-modules":
  [
    /* nb: is ordered list; will be tried top to bottom. */
    {
      "name": "arbitrary string for human identification",
      "class": "parser.UKHASParser",
      "callsigns":
      {
        'a1':
        {
          "config":
          {
            "checksum": "crc16-ccitt",
            "fields":
            [
              {
                "name": "message_count",
                "type": "int"
              },
              /* etc. */
              {
                "name": "internal_temperature",
                "type": "float"
              }
            ]
          },
          "intermediate-filters":
          [
            {
              "type": "normal",
              "callable": "filters.commonhotfixthatsrequiredeverytime"
              "config": { "arbitrary data": True }
            },
            {
              "type": "hotfix",
              "code": "import os\nos.get_root()\nreturn fixed_string",
            }
          ],
          "post-parse-filters":
          [
            /* ... */
          ]
        },
        'habitat':
        {
          /* ... */
        }
      },
      "pre-parse-filters":
      [
        /* pre-parse we don't know what the callsign is so this has to be configured globally to the parser module.
           The filter could try and work out what the callsign is if it has to, but these are for when something is 
           *badly* broken */
      ],
      "wildcard-enable": True,  /* this module is allowed to accept data from any callsign */
      "default-config": { ... } /* config like in callsigns['a1'] for wildcard-enabled parser modules, default when there isn't an entry in callsigns[] */
    }
  ]
}
```

Then (see bullet n. 3) only parser-modules.callsigns[...].config is passed to parse(); since that's all it needs to know. This also means that whatever has to deal with this configuration doesn't care about its contents - the contents are specific to the parser module in use, allowing more flexibility.

**Comment by adamgreig** [comment]
Looks good to me. I'll get on coding it up soonish.

**Comment by adamgreig** [comment]
Okay, I think this might do the trick. It's ready to be reviewed, at any rate.
